File: src/lc_104_max_depth_bt.py

from make_tree import MakeTree

class MaxDepthBinaryTree():

    def __init__(self, tree_array):
        self.tree_array               = tree_array

    def search_max_depth(self, root, sum):

        #Base case
        if root == None:
            return sum
        
        sum += 1

        return max(self.search_max_depth(root.left, sum), 
                   self.search_max_depth(root.right, sum))


    def solve(self):

        #Initialize
        sum     = 0

        mytree  = MakeTree(self.tree_array)

        root    = mytree.get_tree_root()

        return self.search_max_depth(root, sum)

File: src/lc_104_max_depth_bt.c
#include <stdio.h>
#include <make_tree.c>

int max(int a, int b)
{
    if (a > b)
    {
        return a;
    }
    else
    {
        return b;
    }
}

int search_max_depth(struct Node* root)
{
    //Base case
    if (root == NULL)
    {
        return 0;
    }

    //Call recursively
    int max_left    = search_max_depth(root->left);
    int max_right   = search_max_depth(root->right);

    return max(max_left, max_right) + 1;
}

int solve(int arr[], int array_size)
{
    struct Node* root = get_tree_root(arr, array_size);
    
    return search_max_depth(root);
}